extensions/keccak256/circuit/src/trace.rs

```diff
@@ -1,150 +1,216 @@
 };
 
-impl<SC: StarkGenericConfig> Chip<SC> for KeccakVmChip<Val<SC>>
-where
-    Val<SC>: PrimeField32,
-{
-    fn air(&self) -> AirRef<SC> {
-        Arc::new(self.air)
-    }
+impl<F: PrimeField32, CTX> TraceStep<F, CTX> for KeccakVmStep {
+    fn execute(
+        &mut self,
+        state: VmStateMut<TracingMemory<F>, CTX>,
+        instruction: &Instruction<F>,
+        trace: &mut [F],
+        trace_offset: &mut usize,
+        width: usize,
+    ) -> Result<()> {
+        let &Instruction {
+            opcode,
+            a,
+            b,
+            c,
+            d,
+            e,
+            ..
+        } = instruction;
+        let d = d.as_canonical_u32();
+        let e = e.as_canonical_u32();
+        debug_assert_eq!(opcode, Rv32KeccakOpcode::KECCAK256.global_opcode());
+        debug_assert_eq!(d, RV32_REGISTER_AS);
+        debug_assert_eq!(e, RV32_MEMORY_AS);
 
-    fn generate_air_proof_input(self) -> AirProofInput<SC> {
-        let trace_width = self.trace_width();
-        let records = self.records;
-        let total_num_blocks: usize = records.iter().map(|r| r.input_blocks.len()).sum();
-        let mut states = Vec::with_capacity(total_num_blocks);
-        let mut instruction_blocks = Vec::with_capacity(total_num_blocks);
-        let memory = self.offline_memory.lock().unwrap();
+        let trace = &mut trace[*trace_offset..];
+        let (dst, mut src, mut remaining_len) = {
+            let cols: &mut KeccakVmCols<F> = trace[..width].borrow_mut();
+            cols.instruction.start_timestamp = F::from_canonical_u32(state.memory.timestamp());
 
-        #[derive(Clone)]
-        struct StateDiff {
-            /// hi-byte of pre-state
-            pre_hi: [u8; KECCAK_RATE_U16S],
-            /// hi-byte of post-state
-            post_hi: [u8; KECCAK_RATE_U16S],
-            /// if first block
-            register_reads: Option<[RecordId; KECCAK_REGISTER_READS]>,
-            /// if last block
-            digest_writes: Option<[RecordId; KECCAK_DIGEST_WRITES]>,
-        }
+            let a = a.as_canonical_u32();
+            let b = b.as_canonical_u32();
+            let c = c.as_canonical_u32();
+            let dst = tracing_read(state.memory, d, a, &mut cols.mem_oc.register_aux[0]);
+            let src = tracing_read(state.memory, d, b, &mut cols.mem_oc.register_aux[1]);
+            let len = tracing_read(state.memory, d, c, &mut cols.mem_oc.register_aux[2]);
+            (
+                dst,
+                u32::from_le_bytes(src),
+                u32::from_le_bytes(len) as usize,
+            )
+        };
+
+        // Due to the AIR constraints, the final memory timestamp should be the following:
+        let final_timestamp = state.memory.timestamp()
+            + (remaining_len + KECCAK_ABSORB_READS + KECCAK_DIGEST_WRITES) as u32;
+        let num_blocks = num_keccak_f(remaining_len);
+        let mut hasher = Keccak::v256();
+
+        trace
```

Comment:
note: I think the way this is done makes sense ATM, but we'll have to change it / re-think it pretty carefully for the e4 rewrite